### PR TITLE
Prevent [[SetPrototypeOf]] on all globals

### PR DESCRIPTION
**Pull request by domenic**
HTML already prohibits it for WindowProxy; see https://html.spec.whatwg.org/multipage/browsers.html#windowproxy. However, this does not cover the other globals, which need to have their prototype chain locked down. See #133 for previous work.

/cc @littledan @jswalden

***
[Preview](https://cdn.rawgit.com/heycam/webidl/9a39bab/index.html) | [Diff w/ current ED](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fheycam.github.io%2Fwebidl%2F&doc2=https%3A%2F%2Fcdn.rawgit.com%2Fheycam%2Fwebidl%2F9a39bab%2Findex.html) | [Diff w/ base](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fcdn.rawgit.com%2Fheycam%2Fwebidl%2F7cb900a%2Findex.html&doc2=https%3A%2F%2Fcdn.rawgit.com%2Fheycam%2Fwebidl%2F9a39bab%2Findex.html)

### CODE REVIEW

**Comment by littledan** [comment]
LGTM

**Comment by annevk** [comment]
We should add a note here that user agents do not need to implement it for the `Window` object (or make that object opt-out normatively somehow). We went to some lengths to ensure the `Window` object does not need to be an exotic object and I don't think you can write a test that proves it isn't.

**Comment by annevk** [comment]
Hmm, it's even worse. While writing a patch for HTML to export WindowProxy I noticed that HTML explicitly refers to the Window object as an ordinary object. This would make that somewhat contradictory. Shall we just embrace the fact that IDL needs to know about the Window object? (We have some Window<>WindowProxy bridging still to be done I think.)

**Comment by domenic** [comment]
Sure, so omit the [PrimaryGlobal] clause I guess?

**Comment by annevk** [comment]
Well, I think we also want to get rid of [PrimaryGlobal] at some point (and require explicit [Exposed] usage everywhere) so I think it's best to just call out `Window` directly.

**Comment by bzbarsky** [comment]
What about other things on the global's proto chain?  Are those already handled somewhere?

**Comment by bzbarsky** [comment]
Also, is there a reason we're not using https://tc39.github.io/ecma262/#sec-immutable-prototype-exotic-objects here instead of defining a manual override?

**Comment by bzbarsky** [comment]
And in particular a manual override with different behavior?

**Comment by bzbarsky** [comment]
Oh, I guess the proto chain was handled in #133 and in fact used <https://tc39.github.io/ecma262/#sec-immutable-prototype-exotic-objects> (though in a slightly odd way in the case of the named properties object; I don't know why we didn't just make it an immutable prototype exotic object too).

**Comment by domenic** [comment]
An immutable prototype exotic object is, as far as I can tell, an object whose *only* overridden internal method is [[SetPrototypeOf]].

**Comment by bzbarsky** [comment]
@domenic Where does it say that?  It says that an immutable prototype exotic object must have that internal method overridden, but doesn't say anything about the other methods.

**Comment by domenic** [comment]
@bzbarsky: https://tc39.github.io/ecma262/#sec-built-in-exotic-object-internal-methods-and-slots

> The following exotic objects use the ordinary object internal methods except where it is explicitly specified otherwise below:

**Comment by bzbarsky** [comment]
Oh, I see.  Gotta love non-local effects in specs.  :(

**Comment by annevk** [comment]
(So #133 doesn't change the ordinary object status of the Window object since it only affects its prototype object.)

**Comment by bzbarsky** [comment]
OK.  So taking a step back for a second: from my point of view as an implementor, Window is not special in terms of wanting it to be "ordinary".  I want that for all globals.

Which means that my definition of "ordinary" includes immutable prototype exotic objects, given that they only have [[SetPrototypeOf]] modified.  In practice we implement that via a flag on such objects and checking that flag in the default [[SetPrototypeOf]]...

What this should translate to in spec terms, I'm not sure.  But my point was that there current ordinariness of Window is not really sacred, or is no more sacred than the ordinariness of any other global object.

**Comment by domenic** [comment]
So good point on the wrong behavior. Let me push a commit fixing that.

I'm not really concerned about making Window a perfectly ordinary object. I think the note suffices.

**Comment by annevk** [comment]
Will you change HTML too then?

**Comment by domenic** [comment]
I don't really see the need. Window is still an ordinary object in practice since this is unobservable.

**Comment by domenic** [comment]
What is missing to get this merged?

File: index.bs [hunk 1]
```diff
@@ -12261,6 +12252,20 @@ defined in section [[#platformobjectset]].
     1.  Perform steps 3-11 of the [=default [[Set]] internal method=].
 </div>
 
+<h4 id="platformobjectsetprototypeof">Platform object \[[SetPrototypeOf]] method</h4>
+
+<div algorithm="to invoke the internal [[SetPrototypeOf]] method of platform objects">
+
+    The internal \[[SetPrototypeOf]] method of every [=platform object=]
+    that implements an [=interface=] with the [{{Global}}] or [{{PrimaryGlobal}}]
+    [=extended attribute=] must execute the same algorithm must be executed as is defined for the
```
Comment (bzbarsky) [inline index.bs]: "must execute the same algorithm must be executed" has too many "musts".  Probably should be:

  ... must execute the same algorithm as is defined for ...

Comment (domenic) [inline index.bs]: Fixed, thanks

**Pull request merged**